### ISSUE TITLE
tf.function traces twice when creating Variables in the first call?

### ISSUE THREAD

**Issue by levuloihust99** [tensorflow/tensorflow#48851]
I have struggled to understand this strange behavior when using `@tf.function`. I have the following code:
```python
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3' # This prevents tensorflow from print WARNING messages
import tensorflow as tf


class Count:
    def __init__(self):
        self.count = tf.Variable(0)

    @tf.function
    def increment(self):
        print("This function is tracing !")
        tf.print("This function is executing !")
        return self.count.assign_add(1)

counter = Count()
counter.increment()
```
According to [this guide](https://www.tensorflow.org/guide/function#tracing), the above code will first build a `tf. Graph` (**tracing**) and then execute the built `tf. Graph`. When running the above code, I get the expected result as below:
```
This function is tracing!
This function is executing!
```
However, if I try to create the variable `self. count` inside the function decorated with `@tf.function` (`increment` in this case), the function gets traced twice:

*Code:*
```python
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3' # This prevents tensorflow from print WARNING messages
import tensorflow as tf


class Count:
    def __init__(self):
        self.count = None

    @tf.function
    def __call__(self):
        print("This function is tracing !")
        tf.print("This function is executing !")
        if self.count is None:
            self.count = tf.Variable(0)
        return self.count.assign_add(1)

counter = Count()
counter()
```
*Output:*
```
This function is tracing!
This function is tracing!
This function is executing!
```

There is another closely related question, which is [this](https://stackoverflow.com/questions/65323657/why-tf-function-traces-layers-twice). But none of the answers give a clear explanation to my question. I am currently using Python 3.8.5 and TensorFlow 2.4.1

**Comment by bhack** [comment]
Can you try with

```pyhton
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3' # This prevents tensorflow from print WARNING messages
import tensorflow as tf


class Count:
    def __init__(self):
        self.count = None

    @tf.function
    def __call__(self):
        print("This function is tracing !")
        tf.print("This function is executing !")
        if self.count is None:
            self.count = 0
        self.count += 1
        return self.count

counter = Count()
print(counter().numpy())
print(counter.__call__.pretty_printed_concrete_signatures())

```

**Comment by bhack** [comment]
For your example it will need to re-valuate `self.count` when you assign. You can check yourself run with commented and uncommented `#self.count = tf.Variable(0)`

```python
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3' # This prevents tensorflow from print WARNING messages
import tensorflow as tf


class Count:
    def __init__(self):
        self.count = None

    @tf.function
    def __call__(self):
        print("This function is tracing !")
        tf.print("This function is executing !")
        if self.count is None:
            print("True")
            #self.count = tf.Variable(0)
        else:
          print("False")
        return 1

counter = Count()
```

**Comment by levuloihust99** [comment]
> This will trace only once but a second counter() call will print always 1.
> This is a known behavior/limit. You can see more details in a similar example at > https://www.tensorflow.org/guide/function#depending_on_python_objects

I have read this guide and I understand why second counter() always print 1, but that's not my question. My question is **why are there 2 tracings with only one call to `__call__`?**, not **why is there only one tracing with 2 calls to `__call__`?**. Your answer seems to be for the latter question.

> For your example it will need to re-valuate `self.count` when you assign.

Since I am asking for **tracing** behavior of `@tf.function`, I do not get what you mean by this explanation. It seems irrelevant.

I ask this question because I had a problem when training my Keras model. Class like `tf.keras.layers.Dense` defers the creation of Variables until the first call to it. And as I mentioned, the creation of Variables inside a function decorated with `@tf.function` cause that function to trace twice. Hence, my Keras model always traces twice in the first training step, which consumes a lot of time when the model is large. Below code is an example code:

**Code**

```python
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3' # This prevents tensorflow from print WARNING messages
import tensorflow as tf

class MyModel(tf.keras.Model):
    def __init__(self, **kwargs):
        super(MyModel, self).__init__(**kwargs)
        self.dense = tf.keras.layers.Dense(10)

    def call(self, inputs):
        return self.dense(inputs)

model = MyModel()

@tf.function
def train_step(sample):
    print("This function is tracing !")
    model(sample)

tensor = tf.random.normal([128])
dataset = tf.data.Dataset.from_tensors(tensor).repeat(16).batch(4)

for epoch in range(2):
    for sample in dataset:
        train_step(sample)

print("-----------------------------------------------")
print(train_step.pretty_printed_concrete_signatures())
```
**Output**
```
This function is tracing !
This function is tracing !
-----------------------------------------------
train_step(sample)
  Args:
    sample: float32 Tensor, shape=(4, 128)
  Returns:
    NoneTensorSpec()

train_step(sample)
  Args:
    sample: float32 Tensor, shape=(4, 128)
  Returns:
    NoneTensorSpec()
```

As you can see, the two-tracings behavior of `@tf.function` results in two `ConcreteFunction` with exactly same input signatures. According to my understand, it should have only one `ConcreteFunction` (one computation graph) for each input signature. Is this behavior expected or not? If it is expected, what is its purpose?

**Comment by bhack** [comment]
I think this example it is different and probably you can explicit build the model or not?
```python
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3' # This prevents tensorflow from print WARNING messages
import tensorflow as tf

class MyModel(tf.keras.Model):
    def __init__(self, **kwargs):
        super(MyModel, self).__init__(**kwargs)
        self.dense = tf.keras.layers.Dense(10)

    def call(self, inputs):
        return self.dense(inputs)

model = MyModel()

@tf.function
def train_step(sample):
    print("This function is tracing !")
    model(sample)

tensor = tf.random.normal([128])
dataset = tf.data.Dataset.from_tensors(tensor).repeat(16).batch(4)
model.build(input_shape=(4, 128))
sample = next(iter(dataset))
train_step(sample)

print("-----------------------------------------------")
print(train_step.pretty_printed_concrete_signatures())
```

```
This function is tracing !
-----------------------------------------------
train_step(sample)
  Args:
    sample: float32 Tensor, shape=(4, 128)
  Returns:
    NoneTensorSpec()
```

**Comment by levuloihust99** [comment]
Must I always build the model's weights in advance to avoid tracing 2 times? Not only the model but also the optimizer creates some Variables when applying gradients for the first time. Explicitly build the model and optimizer looks not good. Is there a way other than building the model and optimizer explicitly?

Again, my question is **why**. I want to understand why the exactly same computation graph is built 2 times, which does not make sense to me. Explicitly build the model and optimizer can solve the problem, but another way without doing this seems to be better.

**Comment by bhack** [comment]
When you create a `tf.varaible` inside a `tf.function` it will re-trigger a retrace. 
It is also why [Variables may only be created once](https://www.tensorflow.org/api_docs/python/tf/function?version=nightly#variables_may_only_be_created_once) and in the same doc section
> In general, it is recommended to create tf.Variables outside of tf.function

So calling `model(sample)` on an not built model it will go to create model variable inside your `tf.function` and this will trigger the retrace.

The minimal example is:

```
import tensorflow as tf

@tf.function
def train_step():
    #the next line will trigger a retrace inside a tf.function
    tf.Variable(0)
train_step()
```

**Comment by levuloihust99** [comment]
Okay, I got it. Thank you @bhack

**Comment by gabriel-vanzandycke** [comment]
When should the `build()` method be called then? Where is the documentation for that method?

Is it okay to call build() for every batch (only the first call would have an effect) ? Or should I keep track of (un)built models?

**Comment by aoot** [comment]
> When should the `build()` method be called then? Where is the documentation for that method?
> 
> Is it okay to call build() for every batch (only the first call would have an effect) ? Or should I keep track of (un)built models?
> 
> ~Additionally, how should the `build()` method be called when they are multiple inputs to the model?~ Answer in the code.

I'm also figuring out the subtleties between them, however, [this section of the guide](https://www.tensorflow.org/guide/intro_to_modules#the_build_step) contrasts building a simple model using the TF vs Keras, and it highlights the additional lifecycle step in Keras (i.e., `build()`).
